wolfcrypt/src/aria.c

```diff
@@ -0,0 +1,242 @@
+
+        if (rv != MC_OK) { WOLFSSL_MSG(MC_GetErrorString(rv)); return BAD_FUNC_ARG; }
+        return 0;
+    }
+
+    static WARN_UNUSED_RESULT WC_INLINE int CheckAriaGcmIvSize(int ivSz) {
+        return (ivSz == GCM_NONCE_MIN_SZ ||
+                ivSz == GCM_NONCE_MID_SZ ||
+                ivSz == GCM_NONCE_MAX_SZ);
+    }
+
+    int wc_AriaGcmSetExtIV(Aria* aria, const byte* iv, word32 ivSz)
+    {
+        int ret = 0;
+
+        if (aria == NULL || iv == NULL || !CheckAriaGcmIvSize((int)ivSz)) {
+            ret = BAD_FUNC_ARG;
+        }
+
+        if (ret == 0) {
+            XMEMCPY((byte*)aria->nonce, iv, ivSz);
+            aria->nonceSz = ivSz;
+        }
+
+        return ret;
+    }
+
+    int wc_AriaGcmSetIV(Aria* aria, word32 ivSz,
+                   const byte* ivFixed, word32 ivFixedSz,
+                   WC_RNG* rng)
+    {
+        int ret = 0;
+
+        if (aria == NULL || rng == NULL || !CheckAriaGcmIvSize((int)ivSz) ||
+            (ivFixed == NULL && ivFixedSz != 0) ||
+            (ivFixed != NULL && ivFixedSz != AES_IV_FIXED_SZ)) {
+            ret = BAD_FUNC_ARG;
+        }
+
+        if (ret == 0) {
+            byte* iv = (byte*)aria->nonce;
+
+            if (ivFixedSz)
+                XMEMCPY(iv, ivFixed, ivFixedSz);
+
+            ret = wc_RNG_GenerateBlock(rng, iv + ivFixedSz, ivSz - ivFixedSz);
+        }
+
+        if (ret == 0) {
+            aria->nonceSz = ivSz;
+        }
+
+        return ret;
+    }
+
+    int wc_AriaEncrypt(Aria *aria, byte* out, byte* in, word32 inSz,
+                            byte* iv, word32 ivSz, byte* aad, word32 aadSz,
+                            byte* authTag, word32 authTagSz)
+    {
+        MC_RV rv = MC_OK;
+
+        MC_ALGPARAM param = {
+            .pNonce = iv,
+            .nNonce = ivSz,
+            .pAData = aad,
+            .nAData = aadSz,
+            .nTLen = authTagSz,
+            .nDataLen = inSz
+        };
+        MC_ALGORITHM mcAlg = {.mcAlgId = aria->algo, .pParam = (MC_UCHAR*)&param, .nParam = sizeof(param)};
+        MC_UINT outSz = inSz + authTagSz;
+
+        if (authTag == NULL || iv == NULL || authTagSz > ARIA_BLOCK_SIZE ||
+                    authTagSz < WOLFSSL_MIN_AUTH_TAG_SZ ||
+                    ivSz == 0 || ivSz > ARIA_BLOCK_SIZE) {
+            return BAD_FUNC_ARG;
+        }
+        if (out == NULL || in == NULL || inSz == 0) {
+            return BAD_FUNC_ARG;
+        }
+        if (out <= in + inSz && in <= out + inSz) { /* check for overlapping range */
+            return BAD_FUNC_ARG;
+        }
+
+        if (rv == MC_OK) rv = MC_EncryptInit(aria->hSession, &mcAlg, aria->hKey);
+
+        if (rv == MC_OK) rv = MC_Encrypt(aria->hSession, in, inSz, out, &outSz);
+
+        if (rv == MC_OK) XMEMCPY(authTag, out + inSz, authTagSz);
+
+        if (rv != MC_OK) { WOLFSSL_MSG(MC_GetErrorString(rv)); return ENCRYPT_ERROR; }
+        return 0;
+    }
+
+    int wc_AriaDecrypt(Aria *aria, byte* out, byte* in, word32 inSz,
```

Comment:
Please go through and add comments to these new functions. When in the source code here it's great if it has comments for arguments, but at the minimum needs to specify what the return code is going to be, i.e. `ARIA decrypt function which returns 0 on success`.